File: Programmers/codingTest_Kit/Graph/49191.cpp
#include <string>
#include <vector>
#include <iostream>
#define MAX 101
#define INF 9876543231
using namespace std;

int MATRIX[MAX][MAX];

void floydWarshall(int n){
        for(int k = 1; k <= n; k++){
        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= n; j++) 
            MATRIX[i][j] = min(MATRIX[i][j], MATRIX[i][k] + MATRIX[k][j]);
        }
    }
}


int solution(int n, vector<vector<int>> results) {
    int answer = 0;
    for(int i = 1; i <= n; i++){
        for(int j = 1; j <= n; j++) MATRIX[i][j] = INF;
    }
    for(int i = 0; i < results.size() ; i++){
        MATRIX[results[i][0]][results[i][1]] = 1;
    }
    for(int i = 1; i <= n; i++){
        for(int j = 1; j <= n; j++) cout << MATRIX[i][j] << " ";
        cout << '\n';
    }

    floydWarshall(n);
    
    for(int i = 1; i <= n; i++){
        for(int j = 1; j <= n; j++) cout << MATRIX[i][j] << " ";
        cout << '\n';
    }

    for(int i = 1; i <= n; i++){
        int count = 0;
        for(int j = 1; j <= n; j++){
            if(MATRIX[i][j] != INF || MATRIX[j][i] != INF) count++;
        }
        cout << count << '\n';
        if(count == n - 1) answer++;
    }
    return answer;
}

File: Programmers/codingTest_Kit/Graph/49191.py
from collections import Counter


def solution(n, results):
    answer = 0
    board = [[0] * n for _ in range(n)]
    for p1, p2 in results:
        board[p1 - 1][p2 - 1] = 1
        board[p2 - 1][p1 - 1] = -1

    for k in range(n):
        for i in range(n):
            for j in range(n):
                if board[i][j] == 0:
                    if board[i][k] == 1 and board[k][j] == 1:
                        board[i][j] = 1
                    elif board[i][k] == -1 and board[k][j] == -1:
                        board[i][j] = -1

    for i in range(n):
        if Counter(board[i])[0] == 1:
            answer += 1

    return answer